### PR TITLE
154 prototype external dataset storage

### PR DESCRIPTION
**Pull request by pmslavin**
Adds external storage of "large" datasets to a MongoDB instance.

* "Large" datasets are those exceeding a threshold defined in config
* Datasets are transparently migrated to and from external storage as their size requires
* Bulk insert routines updated to handle external storage
* New mongo-specific tests added and existing tests extended
* Utility module added to identify and migrate existing datasets

### CODE REVIEW

**Review by knoxsp — CHANGES_REQUESTED** [review]
It also looks like 'pylibmc' is missing from install_requires.
Perhaps we should install using requirement.txt instead so there's only one list of dependencies?

File: hydra_base/db/model.py [hunk 1]
```diff
@@ -286,26 +293,53 @@ class Dataset(Base, Inspect, PermissionControlled, AuditMixin):
     hash       = Column(BIGINT(),  nullable=False, unique=True)
     cr_date    = Column(TIMESTAMP(),  nullable=False, server_default=text(u'CURRENT_TIMESTAMP'))
     hidden     = Column(String(1),  nullable=False, server_default=text(u"'N'"))
-    value      = Column('value', Text().with_variant(mysql.LONGTEXT, 'mysql'),  nullable=True)
+    value_ref  = Column('value', Text().with_variant(mysql.LONGTEXT, 'mysql'),  nullable=True)
```
[the remark this answers is not in the document]
Comment (pmslavin) [reply hydra_base/db/model.py:295]: Don't think so, the db column is unchanged although the class attr has a new name.

File: hydra_base/lib/adaptors.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+from bson.objectid import ObjectId
+from pymongo import MongoClient
+from pymongo.errors import ServerSelectionTimeoutError
+
+from hydra_base import config
+
+log = logging.getLogger(__name__)
+
+
+class DatasetAdaptor(ABC):
```
[the remark this answers is not in the document]
Comment (pmslavin) [reply hydra_base/lib/adaptors.py]: This has been reorganised in the latest commits.  Is now a Mongo-specific `MongoDatasetAdapter` in a new package `lib.storage`.

File: hydra_base/lib/datasetmanager.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            if datum.key == self.loc_key:
+                return datum.value
+
+
+    def _get_storage_location_query(self, dataset):
+        if not dataset.id:
+            return
+        qry_txt = f"select `value` from tMetadata where dataset_id = {dataset.id} and `key` = '{self.loc_key}'"
+        try:
+            cols = get_session().execute(qry_txt).one()
```
[the remark this answers is not in the document]
Comment (pmslavin) [reply hydra_base/lib/datasetmanager.py]: Original reason was to test performance and avoid having to import `Metadata` from model.py into the function, but the two text-based queries have been removed as they were unused.

File: hydra_base/lib/datasetmanager.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        except NoResultFound:
+            # The dataset's metadata does not have a location key
+            return
+        return cols[0]
+
+
+    def _delete_storage_location_query(self, dataset):
+        dataset_id = getattr(dataset, "id")
+        qry_txt = f"delete from tMetadata where dataset_id = {dataset_id} and `key` = '{self.loc_key}'"
```
[the remark this answers is not in the document]
Comment (pmslavin) [reply hydra_base/lib/datasetmanager.py]: Removed with the other text-based query.

File: hydra_base/lib/objects.py [hunk 5]
```diff
@@ -52,13 +59,49 @@ def __init__(self, obj_dict={}, parent=None, extras={}):
...
-            #A special case, trying to load a SQLAlchemy object, which is a 'dict' object
+            """
+            The argument is a SQLAlchemy object. This originated from a
+            Class.column query so there was no instance to trigger the
+            value descriptor's __get__ and the external lookup must be
+            performed here.
+            """
             obj = obj_dict._asdict()
+            ref_key = obj.get("value")
+            if ref_key:
```
[the remark this answers is not in the document]
Comment (pmslavin) [reply hydra_base/lib/objects.py]: Changed.

File: hydra_base/lib/objects.py [hunk 6]
```diff
@@ -311,3 +356,18 @@ def get_hash(self, val, metadata):
         data_hash = generate_data_hash(dataset_dict)
 
         return data_hash
+
+
+def get_external_value_by_object_id(oid):
```
[the remark this answers is not in the document]
Comment (pmslavin) [reply hydra_base/lib/objects.py:361]: I've removed this whole function and changed the logic in `JSONObject` above in recent commits.

A normal method in the adapter is now called to get the document.

File: hydra_base/util/storage.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+def export_dataset_to_external_storage(ds_id, db_name=None, collection="storage"):
+    """
+    Place the value of the dataset identified by `ds_id` in external
+    storage, replace the value with an ObjectID reference, and
+    update the dataset metadata to indicate the new location.
+    """
+    mongo_config = get_mongo_config()
+    db_name = db_name if db_name else mongo_config["db_name"]
+    collection = collection if collection else mongo_config["datasets"]
```
[the remark this answers is not in the document]
Comment (pmslavin) [reply hydra_base/util/storage.py:61]: The default argument was a mistake from testing which should have been removed.

The default is now `None` so this selects between either an optional arg or the config default.

File: hydra_base/lib/adaptors.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            log.critical(f"Unable to connect to Mongo server {conntext}: {sste}")
+            raise sste
+
+        self.db = self.client[self.db_name]
+
+    def __del__(self):
+        """ Close connection on object destruction """
+        self.client.close()
+
+    def get_document_by_object_id(self, object_id, collection=None):
```
[the remark this answers is not in the document]
Comment (pmslavin) [reply hydra_base/lib/adaptors.py]: Added.

**Comment by pmslavin** [comment]
This has several changes:
* Adaptor => Adapter as this seems more usual
* External storage-related classes refactored into a `storage` package
*  Adapter does not derive from an abstract base class: because there's adapter-specific behaviour in a `DatasetManager` class that uses it, there's no point having adapters conform to a consistent interface when each manager uses only one adapter
*  `DatasetManager` *does* derive from an abstract base class as this is the entity that changes to swap in a different external storage source, ie we need only set `Dataset.value = SomeDatasetManager()` in `model.py` to swap. This class provides the expected descriptor methods and others which manage the storage location metadata
* A few isolated functions are now included in the adapter class
* CI workflow has been added. It seems the runner Ubuntu image includes (the official -org version of) the MongoDB package, so it is only necessary to start this
*  Fixes and clarifications as per comments
 
There's no mention of `pylibmc` in `install_requires` as there's no caching yet in this branch.  It does now include `pymongo` and the CI workflow succeeds with this.

File: hydra_base/lib/data.py [hunk 9]
```diff
@@ -583,6 +579,34 @@ def _bulk_insert_data(bulk_data, user_id=None, source=None):
...
+    mongo_data = {}
+
+    for idx, ds in enumerate(new_data_for_insert):
+        ds_size = len(ds["value"])
+        if ds_size > threshold_sz:
+            mongo_data[idx] = ds["value"]
+            ds_metadata = metadata[ds["hash"]]
+            ds_metadata[loc_key] = mongo_location_token
+
+    if mongo_data:
```
Comment (knoxsp) [inline hydra_base/lib/data.py:605]:
```suggestion
    if mongo_data is not None:
```

**Review by knoxsp — APPROVED** [review]

**Review by knoxsp — APPROVED** [review]

**Pull request merged**